### PR TITLE
Scott proof new

### PR DESCRIPTION
**Pull request by scottviteri**
First working example of translation to Lean proofs.

In SMTLib:
(set-logic QF_UF)
(set-info :smt-lib-version 2.0)
(set-info :category "crafted")
(set-info :status unsat)
(declare-sort U 0)
(declare-fun a () U)
(declare-fun b () U)
(assert (= a b))
(assert (not (= a b)))
(check-sat)
(exit)

In Lean:
open smt
open smt.sort smt.term
def U := sort.atom 1
def a := const 1 U
def b := const 2 U
noncomputable theorem th0 : holds [(mkEq a b)] -> holds [(mkNot (mkEq a b))] -> holds [] :=
(assume (v0 : holds [(mkNot (mkEq a b))]),
(assume (v1 : holds [(mkEq a b)]),
R0 v1 v0 (mkEq a b)))

### CODE REVIEW

File: src/proof/lean/lean_printer.h [hunk 1]
```diff
@@ -19,30 +19,221 @@
...
+  switch (id)
+  {
+    case LeanRule::TRUST:
+    {
+      out << "trust\n";
+      break;
+    }
+    case LeanRule::ASSUME:
+    {
+      int var_index = passumeMap.size();
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:159]:
```suggestion
      size_t var_index = passumeMap.size();
```

**Review by HanielB — COMMENTED** [review]
This is great, Scott. :) I've added several comments, please take a look. I'm still gonna take a look at the lean_post_processor.

File: src/smt/proof_manager.cpp [hunk 2]
```diff
@@ -119,22 +122,26 @@ void PfManager::setFinalProof(std::shared_ptr<ProofNode> pfn,
...
 }
 
-void PfManager::printProof(std::ostream& out,
-                           std::shared_ptr<ProofNode> pfn,
-                           Assertions& as,
-                           DefinedFunctionMap& df)
+  void PfManager::printProof(std::ostream& out,
+                             std::shared_ptr<ProofNode> pfn,
+                             Assertions& as,
+                             DefinedFunctionMap& df)
```
[the remark this answers is not in the document]
Comment (scottviteri) [reply src/smt/proof_manager.cpp:128]: How can I revert added lines? It's not like I can remove printProof itself

File: src/smt/proof_manager.cpp [hunk 3]
```diff
@@ -119,22 +122,26 @@ void PfManager::setFinalProof(std::shared_ptr<ProofNode> pfn,
...
   // TODO (proj #37) according to the proof format, post process the proof node
   // TODO (proj #37) according to the proof format, print the proof node
   if (options::proofFormatMode() == options::ProofFormatMode::LEAN)
   {
+    std::vector<Node> assertions;
+    getAssertions(as, df, assertions);
+    d_lpfpp.reset(new proof::LeanProofPostprocess(d_pnm.get()));
     d_lpfpp->process(fp);
-    proof::leanPrinter(out, fp);
+    CVC4::proof::leanPrinter(out, assertions, fp);
```
[the remark this answers is not in the document]
Comment (scottviteri) [reply src/smt/proof_manager.cpp:140]: See lp.print(out, assertions, fp.get()); for the LFSC printer.
I use the assertions input to get the names of sorts with expr::getSymbols.

File: src/smt/proof_manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/smt/proof_manager.cpp:128]: I meant the extra spacing you used. The issue here is that your code is not clang-formatted, as it'd fix such an issue.

File: src/smt/proof_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/smt/proof_manager.cpp]: I see, cool.

[the remark this answers is not in the document]
Comment (HanielB) [reply src/smt/proof_manager.cpp]: Actually, I remember now why I wrote this. With "this" I meant the `CVC4::` prefix.

File: src/proof/lean/lean_post_processor.cpp [hunk 4]
```diff
@@ -14,17 +14,19 @@
 
 #include "proof/lean/lean_post_processor.h"
 
+#include "../../expr/proof_node_updater.h"
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:17]: delete

File: src/proof/lean/lean_post_processor.cpp [hunk 5]
```diff
@@ -36,7 +38,7 @@ LeanProofPostprocess::LeanProofPostprocess(ProofNodeManager* pnm)
 bool LeanProofPostprocessCallback::shouldUpdate(std::shared_ptr<ProofNode> pn,
                                                 bool& continueUpdate)
 {
-  return false;
+  return (pn->getRule() != PfRule::LEAN_RULE);
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:41]:
```suggestion
  return pn->getRule() != PfRule::LEAN_RULE;
```

File: src/proof/lean/lean_post_processor.cpp [hunk 6]
```diff
@@ -46,9 +48,183 @@ bool LeanProofPostprocessCallback::update(Node res,
                                           CDProof* cdp,
                                           bool& continueUpdate)
 {
+  NodeManager* nm = NodeManager::currentNM();
+  // change to case
+  // Trace("Hello") << id << "\n";
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:53]: delete

File: src/proof/lean/lean_post_processor.cpp [hunk 7]
```diff
@@ -46,9 +48,183 @@ bool LeanProofPostprocessCallback::update(Node res,
...
+  switch (id)
+  {
+    case PfRule::REFL:
+    {
+      Node lean_id =
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:58]:
```suggestion
      Node leanId =
```
Use caml case

File: src/proof/lean/lean_post_processor.cpp [hunk 8]
```diff
@@ -46,9 +48,183 @@ bool LeanProofPostprocessCallback::update(Node res,
...
+          std::vector<Node> lean_args = {lean_id, res, newArgs[1]};
+          cdp->addStep(cur, PfRule::LEAN_RULE, newChildren, lean_args);
+        }
+        else
+        {
+          Node lean_id =
+              nm->mkConst<Rational>(static_cast<unsigned>(LeanRule::R0));
+          std::vector<Node> lean_args = {lean_id, res, newArgs[1]};
+          cdp->addStep(cur, PfRule::LEAN_RULE, newChildren, lean_args);
+        }
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:105]:
```suggestion
        Node leanId = nm->mkConst<Rational>(static_cast<unsigned>(
            newArgs[0].getConst<bool>() ? LeanRule::R0 : LeanRule::R1));
        std::vector<Node> lean_args = {leanId, res, newArgs[1]};
        cdp->addStep(cur, PfRule::LEAN_RULE, newChildren, lean_args);
```

File: src/proof/lean/lean_post_processor.cpp [hunk 9]
```diff
@@ -46,9 +48,183 @@ bool LeanProofPostprocessCallback::update(Node res,
...
+    {
+      Node lean_id =
+          nm->mkConst<Rational>(static_cast<unsigned>(LeanRule::ASSUME));
+      std::vector<Node> lean_args = {lean_id, res};
+      lean_args.insert(lean_args.end(), args.begin(), args.end());
+      cdp->addStep(res, PfRule::LEAN_RULE, children, lean_args);
+      break;
+    }
+
+    /*
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:119]: delete commented out code in general

File: src/proof/lean/lean_post_processor.cpp [hunk 10]
```diff
@@ -36,19 +38,103 @@ LeanProofPostprocess::LeanProofPostprocess(ProofNodeManager* pnm)
...
+      else
+      {
+        new_id = d_nm->mkConst<Rational>(
+            static_cast<unsigned>(proof::LeanRule::SMTSYMM_NEG));
+        t1 = child[0][0];
+        t2 = child[0][1];
+        c1 = d_nm->mkNode(kind::EQUAL, t1, t2);
+        c2 = d_nm->mkNode(kind::NOT, d_nm->mkNode(kind::EQUAL, t2, t1));
+        new_res = d_nm->mkNode(c1, c2);
+      }
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:118]:
```suggestion
      if (k == kind::EQUAL)
      {
        new_id = d_nm->mkConst<Rational>(
            static_cast<unsigned>(proof::LeanRule::SMTSYMM));
        t1 = child[0];
        t2 = child[1];
        c1 = d_nm->mkNode(kind::NOT, d_nm->mkNode(kind::EQUAL, t1, t2));
        c2 = d_nm->mkNode(kind::EQUAL, t2, t1);
      }
      else
      {
        new_id = d_nm->mkConst<Rational>(
            static_cast<unsigned>(proof::LeanRule::SMTSYMM_NEG));
        t1 = child[0][0];
        t2 = child[0][1];
        c1 = d_nm->mkNode(kind::EQUAL, t1, t2);
        c2 = d_nm->mkNode(kind::NOT, d_nm->mkNode(kind::EQUAL, t2, t1));
      }
      new_res = d_nm->mkNode(c1, c2);
```

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -154,6 +155,9 @@ libcvc4_add_sources(
   proof/proof_letify.h
   proof/proof_manager.cpp
   proof/proof_manager.h
+  proof/lean/lean_post_processor.cpp
+  proof/lean/lean_post_processor.h
+  proof/lean/lean_printer.h
```
Comment (HanielB) [inline src/CMakeLists.txt:160]: delete this

File: src/smt/proof_manager.cpp [hunk 12]
```diff
@@ -37,8 +37,8 @@ PfManager::PfManager(context::UserContext* u, SmtEngine* smte)
       d_rewriteDb(new theory::RewriteDb),
       d_pppg(new PreprocessProofGenerator(
           d_pnm.get(), u, "smt::PreprocessProofGenerator")),
-      d_pfpp(new ProofPostproccess(d_pnm.get(), smte, d_pppg.get())),
-      d_lpfpp(new proof::LeanProofPostprocess(d_pnm.get())),
+      d_pfpp(new ProofPostprocess(d_pnm.get(), smte, d_pppg.get())),
```
Comment (HanielB) [inline src/smt/proof_manager.cpp:40]: revert this change

File: src/smt/proof_manager.h [hunk 13]
```diff
@@ -115,9 +115,10 @@ class PfManager
   /** The preprocess proof generator. */
   std::unique_ptr<smt::PreprocessProofGenerator> d_pppg;
   /** The proof post-processor */
-  std::unique_ptr<smt::ProofPostproccess> d_pfpp;
+  std::unique_ptr<smt::ProofPostprocess> d_pfpp;
```
Comment (HanielB) [inline src/smt/proof_manager.h:118]: revert this change

**Review by HanielB — COMMENTED** [review]
Revert all the changes you did to `smt/proof_post_processor.h` and `smt/proof_post_processor.cpp`

File: src/CMakeLists.txt (hunk 11 above)
Comment (scottviteri) [reply src/CMakeLists.txt:160]: ok

File: src/proof/lean/lean_post_processor.cpp [hunk 14]
```diff
@@ -36,19 +38,85 @@ LeanProofPostprocess::LeanProofPostprocess(ProofNodeManager* pnm)
...
+                  *cdp);
+      break;
+    }
+    default:
+    {
+      return false;
+    }
+  };
   return true;
+}  // namespace proof
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.cpp:113]:
```suggestion
}
```

File: src/proof/lean/lean_printer.cpp [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+{
+  uint32_t id;
+  if (ProofRuleChecker::getUInt32(n, id))
+  {
+    return static_cast<LeanRule>(id);
+  }
+  return LeanRule::UNKNOWN;
+}
+
+  //Node LeanPrinter::getId(std::shared_ptr<ProofNode> n)
```
Comment (HanielB) [inline src/proof/lean/lean_printer.cpp:61]: remove all commented out code

File: src/proof/lean/lean_printer.h [hunk 16]
```diff
@@ -19,31 +19,78 @@
...
 
-enum class LeanRule : uint32_t
-{
-  // in what format should I put these lean rules
-  // all the lean rules
-};
+/**
+ * Converts a lean rule to a string.
+ * @param id The lfsc rule
+ * @return The name of the lfsc rule
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:34]: fix

File: src/proof/lean/lean_printer.h [hunk 17]
```diff
@@ -19,31 +19,78 @@
...
+   */
+  static void printLeanType(std::ostream& s, Node n);
+  /**
+   * Print Lean type corresponding to proof of unsatisfiability
+   */
+  static void printLeanTypeToBottom(std::ostream& s, Node n);
+  /**
+   * Print user defined sorts and constants of those sorts
+   */
+  static void printSorts(std::ostream& out,
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:84]: change the name to say you're printing not only sorts but also constants

File: src/proof/lean/lean_printer.h [hunk 18]
```diff
@@ -19,31 +19,78 @@
...
+   * Convert a node to a Lean term -- must start with mk_ and take children as args
+   * Example: kind::AND (kind::EQUAL a b) c --> mkAnd (mkEq a b) c
+   */
+  static void printLeanString(std::ostream& s, Node n);
+  /**
+   * Convert from node to lean type syntax
+   */
+  static void printLeanType(std::ostream& s, Node n);
+  /**
+   * Print Lean type corresponding to proof of unsatisfiability
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:78]: say why "typetobottom".

File: src/proof/lean/lean_printer.h [hunk 19]
```diff
@@ -19,31 +19,78 @@
...
+  //static Node getId(std::shared_ptr<ProofNode> n);
+  //static Node getConclusion(std::shared_ptr<ProofNode> n);
+  static void printKind(std::ostream& s, Kind k);
+  /**
+   * Convert a node to a Lean term -- must start with mk_ and take children as args
+   * Example: kind::AND (kind::EQUAL a b) c --> mkAnd (mkEq a b) c
+   */
+  static void printLeanString(std::ostream& s, Node n);
+  /**
+   * Convert from node to lean type syntax
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:74]: Use "Lean" rather than "lean" everywhere. Clarify why you talk about converting a type but receives a Node as input.

File: src/proof/lean/lean_printer.h [hunk 20]
```diff
@@ -19,31 +19,78 @@
...
+                    const std::vector<Node>& assertions,
+                    std::shared_ptr<ProofNode> pfn);
+
+ private:
+  /**
+   * Convert a CVC4 Node holding an id to the corresponding LeanRule
+   */
+  static LeanRule getLeanRule(Node n);
+  //static Node getId(std::shared_ptr<ProofNode> n);
+  //static Node getConclusion(std::shared_ptr<ProofNode> n);
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:66]: delete

File: src/proof/lean/lean_printer.h [hunk 21]
```diff
@@ -19,31 +19,78 @@
...
+  static void printKind(std::ostream& s, Kind k);
+  /**
+   * Convert a node to a Lean term -- must start with mk_ and take children as args
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:69]: Say why it must start like this.

File: src/proof/lean/lean_printer.h [hunk 22]
```diff
@@ -19,31 +19,78 @@
...
+   */
+  static void printLeanTypeToBottom(std::ostream& s, Node n);
+  /**
+   * Print user defined sorts and constants of those sorts
+   */
+  static void printSorts(std::ostream& out,
+                         const std::vector<Node>& assertions,
+                         std::shared_ptr<ProofNode> pfn);
+  /**
+   * Print rule specific lean syntax, traversing children before parents in ProofNode tree.
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:88]: Rephrase. Also clarify why the traversal is like this.

File: src/proof/lean/lean_post_processor.h [hunk 23]
```diff
@@ -52,6 +54,12 @@ class LeanProofPostprocessCallback : public ProofNodeUpdaterCallback
  private:
   /** The proof node manager */
   ProofNodeManager* d_pnm;
+  ProofChecker* d_pc;
+  bool addLeanStep(Node res,
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.h:58]: document

File: src/proof/lean/lean_post_processor.h [hunk 24]
```diff
@@ -35,8 +37,8 @@ class LeanProofPostprocessCallback : public ProofNodeUpdaterCallback
  public:
   LeanProofPostprocessCallback(ProofNodeManager* pnm);
   /**
-   * Initialize, called once for each new ProofNode to process. This initializes
-   * static information to be used by successive calls to update.
+   * Initialize, called once for each new ProofNode to process. This
+   * initializes static information to be used by successive calls to update.
    */
   void initializeUpdate();
   bool shouldUpdate(std::shared_ptr<ProofNode> pn,
```
Comment (HanielB) [inline src/proof/lean/lean_post_processor.h:44]: document

**Review by HanielB — COMMENTED** [review]
More comments. Please be careful to address all outstanding comments before doing other things.

File: src/proof/lean/lean_printer.h [hunk 25]
```diff
@@ -62,8 +62,13 @@ class LeanPrinter
    * Convert a CVC4 Node holding an id to the corresponding LeanRule
    */
   static LeanRule getLeanRule(Node n);
-  //static Node getId(std::shared_ptr<ProofNode> n);
-  //static Node getConclusion(std::shared_ptr<ProofNode> n);
+  /**
+   * The Lean calculus uses rules such as mkEq, which wraps (eq x y)
+   *  with a check that the x and y have the same sort.
+   * printKind cases of proof rule kinds such as equality,
+   *  logical and, or logical not, and translating to mkEq,
```
Comment (HanielB) [inline src/proof/lean/lean_printer.h:69]: Improve the writing, hard to parse currently.
